### PR TITLE
kernel: make AppId unique

### PR DESCRIPTION
**Pull request by bradjc**
`AppId.idx` used to refer to the index of the app in the process array. However, if an app restarts then an "old" AppId that was created for the process before it restarted will also work after an app is restarted, but the restarted version may use entirely different resources. This changes the AppId after an app is restarted so that to the rest of the OS it looks like an entirely different application.

Rather than passing around `idx` we now pass the entire `AppId` to make the interface more clear.

Requires #1533

### Testing Strategy

This pull request was tested by working on app restarting.

### TODO or Help Wanted

I think this is a reasonable change that improves this abstraction in the kernel. But, there may be a more Rusty method to accomplish the goal of disabling AppIds after the app crashes/is removed/restarts.

~~This breaks IPC currently since that relies on idx being an actual index in an array.~~

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by alevy** [comment]
This passes the tests against master now that #1533 is merged, maybe rebase and push to pass the other tests?

File: kernel/src/sched.rs [hunk 1]
```diff
@@ -29,6 +29,10 @@ pub struct Kernel {
     work: Cell<usize>,
     /// This holds a pointer to the static array of Process pointers.
     processes: &'static [Option<&'static dyn process::ProcessType>],
+    /// Value to keep track of how many process identifiers have been used. Each
+    /// new process will get the next process identifier and this will be
+    /// incremented.
+    process_identifier_max: Cell<usize>,
```
Comment (alevy) [inline kernel/src/sched.rs]: Maybe this should explicitly be a u64? `usize` is mostly going to be u32, and that's small enough to maybe overflow with enough process restarts.

Comment (bradjc) [reply kernel/src/sched.rs]: I can do that.

**Comment by bradjc** [comment]
There is a bug in this as iterating through grant regions needs to be updated.

**Comment by bradjc** [comment]
Ok grants now iterate the processes array rather than rely on app ids to match the array indices.

**Comment by bradjc** [comment]
There are really two changes in this PR:

1. Remove the assumption that `AppId.idx` is equivalent to the index in the `PROCESSES` array where that process's state is stored.
2. Choose a unique "app_identifier" number when a process is started or restarted.

I think (1) is a reasonable change that provides flexibility going forward, and I don't think we gain much by having `AppId.idx` corresponding to the index in the array directly.

Number (2), however, is a policy change, and while it does help with restarting the ADC capsule as it is currently written, it might not be the right change to make. The goal of creating a unique number is that if a capsule (or anything) has a copy of an `AppId`, and the process that appid number corresponds to goes away for whatever reason, any use of that `AppId` will result in an error since that number no longer matches a valid process. Grants behave in a similar way, where if a capsule has a reference to a grant and the process goes away, the grant will either be re-created, skipped (in case of the `.each()` call), or an error will be returned. After thinking about this more the grant approach is quite nice, and it might make sense to mirror that with `AppId`.

I think I will split this into two PRs to make it easier to discuss both aspects.

**Pull request closed without merging**